### DISCUSSION TITLE
maxEntries and maxAgeSeconds not working

### DISCUSSION THREAD

**Discussion by Viktor-as** [DuCanhGH/next-pwa#103] [Q&A]
Hi, I need help 🙏 

I am caching images and I want the cache to delete it self every day and not to cache more then for example 50 images.

This is my next.config.js:

```
/** @type {import('next').NextConfig} */
const { i18n } = require('./next-i18next.config');
// @ts-check
const withTranslateRoutes = require('next-translate-routes/plugin');
const runtimeCaching = require('./services/cache');

const withPWA = require('@ducanh2912/next-pwa').default({
  dest: 'public',
  register: true,
  skipWaiting: true,
  fallbacks: {
    image: '/offline/offline.png',
  },
  disable: process.env.NODE_ENV === 'development',
  workboxOptions: {
    runtimeCaching: runtimeCaching,
    cleanupOutdatedCaches: true,
  },
  cacheOnFrontEndNav: false,
  reloadOnOnline: true,
});

const nextConfig = withPWA(
  withTranslateRoutes({
    webpack: (config) => {
      config.module.rules.push({
        test: /\.svg$/,
        use: [{ loader: '@svgr/webpack', options: { icon: true } }],
      });
      return config;
    },
    pageExtensions: ['tsx', 'js'],
    i18n,
    reactStrictMode: false,
  })
);
module.exports = nextConfig;
```

I import my runtimeCaching rules from another file ./services/cache.js it has this code:

```
module.exports = [
{
    urlPattern: ({ request, url, sameOrigin }) => {
      //other images
      if (!sameOrigin) return false;
      if (url.pathname === '/_next/image') {
        // console.log('url-other-images', url.href);
        return true;
      }
      return false;
    },
    handler: 'NetworkFirst',
    options: {
      cacheName: 'other-images',
      expiration: {
        maxEntries: 20,
        maxAgeSeconds: 60, // 1 min
      },
    },
  },
]
```

So I am able to cache all images, but maxEntries: 20 and maxAgeSeconds: 60, does not work. In chrome dev tools I see in the cache like 100+ images, and they do not disappear after 1 min. My PWA app gets very big, very fast.

Maybe I am configuring something wrong?

**Comment by DuCanhGH** [comment]
Well if `workboxOptions.*` options do not work, you should open an issue in `GoogleChrome/workbox` rather than this one unless you are certain that this is due to `next-pwa` modifying the options, which doesn't seem to be the case in this one.

That said, I found this in Workbox's documentation:

> Because it's slow to open IndexedDB, expiration won't occur until after the request is used. This means that an expired request may be used once, but will be expired after that. To alleviate this, the plugin will check the Date header of the cached response, if one exists and the date can be parsed, it'll expire based on this as it doesn't require an IndexedDB lookup.

Are you sure that you had requested the files again after they expired?

Just a side note, but `skipWaiting` is not a valid `next-pwa` option, although `workboxOptions.skipWaiting` is. You should consider adding `// @ts-check` to the top of your `next.config.js` to leverage type checking.

**Reply by Viktor-as** [reply]
Thanks @DuCanhGH for your help!

I am not certain where this issue is coming from :/

Yes I requested the files again, but I can still see them in cache. Well know that you mentioned I noticed that the "Time cached" updates. I do not understand then the purpose of this.... If I not request the files they just sit in the cache after expiration, and if I request then they just update the time cached entry.....

I still do not know if its a bug or just I am doing something wrong 🤔 

Thanks for side notes, I fixed them

**Reply by DuCanhGH** [reply]
@Viktor-as I found this explanation from an ex-maintainer of `GoogleChrome/workbox`:

> When using workbox-expiration, cleanup happens after a request is read/written to the cache.
> So what would happen if your web app starts up after a period of inactivity is that after the first request is made for something that matches that route, Workbox would check all entries in the cache, delete the least-recently used ones if there are more than 50, and also delete any that were used more than 1 day ago.

There's also this in the documentation:

> Whenever you update a cached entry, you need to call the updateTimestamp() method so that its age is updated.
> ```js
> await openCache.put(request, response);
> await expirationManager.updateTimestamp(request.url);
> ```

Seems to well explain the mechanism behind `workbox-expiration` :) I believe it means that the timestamp updates are expected.

Your initial comment mentions that there were like 100+ cache entries, so perhaps it's a bug. I'd open a issue in `GoogleChrome/workbox`, but it's fairly dead right now to be honest, so it's a shame.

**Reply by Viktor-as** [reply]
Yeah, the timestamp updates are expected and they are working, but this part is not working (at least for me): 

> Workbox would check all entries in the cache, delete the least-recently used ones if there are more than 50, and also delete any that were used more than 1 day ago.

I figured a way to delete the cache manually every day on home page load, so I will share if someone will need in the future

```jsx
// Manuall cache delete after 1 day
  useEffect(() => {
    async function deleteCache() {
      const imageCache = await caches.open('other-images');
      const imageCacheKeys = await imageCache.keys();
      const cachedResponsesImg = await imageCache.matchAll();

      if (cachedResponsesImg) {
        const currentTime = Date.now();

        cachedResponsesImg.forEach(async (cachedResponse, index) => {
          const cachedTimestamp = new Date(
            cachedResponse.headers.get('date')!
          ).getTime();

          if ((currentTime - cachedTimestamp) / 1000 > 60 * 60 * 24) {
            await imageCache.delete(imageCacheKeys[index]);
          }
        });
      }
    }
    deleteCache();
  }, []);
```

**Reply by DuCanhGH** [reply]
@Viktor-as cool! Suggest you open an issue in their repo as well.

We also allow you to write custom worker code, so just move this thing to `worker/index.ts` and let `next-pwa` handle it :) No need to have an `useEffect`.

**Reply by Viktor-as** [reply]
Hey @DuCanhGH 

I opened an issue at `GoogleChrome/workbox`

Would love to just move this thing to `worker/index.ts` but I am a bit confused on how to do that 🤔 

I should add my code to an event listener? but which? If I add my code to `install` event listener it would run only when I deploy a new build?

So should my code look something like this in `worker/index.ts` ?:

```ts
declare const self: ServiceWorkerGlobalScope;

self.addEventListener('fetch', (event) => {
    event.waitUntil(
      (async () => {
        const imageCache = await caches.open('other-images');
        const imageCacheKeys = await imageCache.keys();
        const cachedResponsesImg = await imageCache.matchAll();

        if (cachedResponsesImg) {
          const currentTime = Date.now();

          cachedResponsesImg.forEach(async (cachedResponse, index) => {
            const cachedTimestamp = new Date(
              cachedResponse.headers.get('date')!
            ).getTime();

            if ((currentTime - cachedTimestamp) / 1000 > 60 * 60 * 24) {
              await imageCache.delete(imageCacheKeys[index]);
            }
          });
        }
      })()
    );
  });

```
Also I get typescript errors on line `declare const self: ServiceWorkerGlobalScope;`
Cannot redeclare block-scoped variable 'self'.ts(2451) lib.dom.d.ts(27317, 13): 'self' was also declared here.
Cannot find name 'ServiceWorkerGlobalScope'.ts(2304)

**Reply by DuCanhGH** [reply]
@Viktor-as well if you want to clean the image cache up on every request (even those that are not images, so you should check if the request is for an image using a RegExp, preferably the same one as what you used in `workboxOptions.runtimeCaching`) then it is probably like that. But the equivalent of the code with `useEffect` should be something like:
```ts
declare const self: ServiceWorkerGlobalScope;

(async () => {
  const imageCache = await caches.open("other-images");
  const imageCacheKeys = await imageCache.keys();
  const cachedResponsesImage = await imageCache.matchAll();

  if (cachedResponsesImage) {
    const currentTime = Date.now();

    cachedResponsesImage.forEach(async (cachedResponse, index) => {
      const cachedTimestamp = new Date(
        cachedResponse.headers.get("date")!
      ).getTime();

      if ((currentTime - cachedTimestamp) / 1000 > 60 * 60 * 24) {
        await imageCache.delete(imageCacheKeys[index]);
      }
    });
  }
})();

export {};
```

That TS error can be fixed by adding `"webworker"` to `compilerOptions.lib` in `tsconfig.json` and marking the file as a module by either adding `export {}` or importing something.

**Reply by Viktor-as** [reply]
Probably not the best idea to run on every fetch :D

Your advice fixed the TS error. But when should now this code run? As I am not able to trigger the image cache deletion after I moved the code into  `worker/index.ts` 

It looks that it is triggered on new build, when a new service worker is installed, but after that the code does not get triggered again 🤔
I tried refreshing the page and opening in new tab. Changed `(currentTime - cachedTimestamp) / 1000 > 60` to delete img cache older then 1 min

**Reply by DuCanhGH** [reply]
@Viktor-as sorry for forgetting to reply 💀 The notification is marked as read so I thought there were no issues on your side.

That the images only get deleted ~~on page load~~ is expected since `worker/index.ts` is only run once. ~~If that isn't the case then I don't really know either because that's just weird.~~

You can use a RegExp to check if the request's URL is pointing towards an image, in which case, run the cleanup code. Should work well enough while you wait for a response from Workbox (tbh, that won't come anytime soon).

**Reply by Viktor-as** [reply]
Its ok @DuCanhGH no worries.

Well I put some `conole.log` in your code example in `worker/index.ts` then I get those console logs printed only when I rebuild the website and on first load. I guess this triggers only when a new service worker is installing. But not on the following site reloads.

Yeah I could add the cache cleaning logic to fetch event listener and with RegExp  narrow down to only image requests, but would this give me any advantages? For example in my homepage I have like 10 images, so the code would run 10 times. 

For now I left the cleanup code in useEffect in my homepage to run only once on page load.

And yeah, I am not really hoping for workbox to fix this issue or even answer  ☠️

**Reply by DuCanhGH** [reply]
@Viktor-as ah yes, my bad, `worker/index.ts` is only executed once on each service worker installation. Not really in the best state of mind today. I mean, I straight up went and consecutively pushed several commits to `master` today. 

With this in mind, perhaps using `useEffect` makes more sense than that. I'd still say that you can just listen to `"fetch"` event and (use `Promise.all` to) delete all the expired files (concurrently) though. This ensures that the files are always cleaned up properly. I mean, this is `workbox-expiration`'s code: 
https://github.com/GoogleChrome/workbox/blob/v7/packages/workbox-expiration/src/CacheExpiration.ts#L99-L144 https://github.com/GoogleChrome/workbox/blob/v7/packages/workbox-expiration/src/ExpirationPlugin.ts#L145-L183

Not that different to be honest. They have this `dontWaitFor` which is kinda crazy to me though. Perhaps you can consider throwing that stuff into your code as well 💀 I mean, that's probably to shut up their linter I guess.

**Reply by Viktor-as** [reply]
>  I mean, I straight up went and consecutively pushed several commits to master today.

@DuCanhGH hehe, happens :D

Well I guess I leave it in useEffect for now as I moved already to other tasks. If I come back to PWA task I will try to listen on `fetch` events and do the cleaning.

Thanks again for the help, and good luck :)
